src/main/java/com/capitalone/dashboard/controller/SonarController.java

```diff
@@ -0,0 +1,145 @@
+    }
+
+    @RequestMapping(value = "/refresh", method = GET, produces = APPLICATION_JSON_VALUE)
+    public ResponseEntity<String> refresh(@Valid String projectName, @Valid String projectKey, @Valid String instanceUrl) {
+
+        gatherParams(projectName, projectKey, instanceUrl);
+        if (Objects.nonNull(instanceUrl)) {
+            this.sonarClient = this.sonarClientSelector.getSonarClient(this.sonarClientSelector.getSonarVersion(instanceUrl));
+            this.collector = collectorRepository.findByName("Sonar");
+
+            SonarProject projectToRefresh;
+            if (Objects.nonNull(this.collector)) {
+                if ((Objects.nonNull(projectName) && Objects.nonNull(projectToRefresh = getExistingProject()))
+                || (Objects.nonNull(projectKey) && Objects.nonNull(projectToRefresh = createNewProjectIfNotExists()))) {
+                    updateCodeQualityData(projectToRefresh);
+                    return response("successfully refreshed sonar project");
+                }
+                return response("unable to refresh sonar project");
+            }
+            return response("sonar collector not found");
+        }
+        return response("sonar instance url is invalid");
+    }
+
+    private void gatherParams(String projectName, String projectKey, String instanceUrl) {
+        this.projectName = projectName;
+        this.projectKey = projectKey;
+        this.instanceUrl = instanceUrl;
+    }
+
+    private ResponseEntity<String> response(String message) {
+        return ResponseEntity.status(HttpStatus.OK)
+                .body(String.format(message + " - projectName=%s projectKey=%s instanceUrl=%s ",
+                        Objects.toString(this.projectName, ""), Objects.toString(this.projectKey, ""), this.instanceUrl));
+    }
+
+    private SonarProject createNewProjectIfNotExists() {
+        SonarProject project = null;
+        if (Objects.nonNull(this.projectKey)) {
+            SonarProject newProject = this.sonarClient.getProject(this.projectKey, this.instanceUrl);
+            if (Objects.nonNull(newProject)) {
+                SonarProject existingProject = sonarProjectRepository.findSonarProject(this.collector.getId(),
+                        newProject.getInstanceUrl(), newProject.getProjectId());
+                project = Objects.nonNull(existingProject) ? existingProject : newProject;
+            }
+        }
+        return project;
+    }
+
+    private SonarProject getExistingProject() {
+        List<SonarProject> sonarProjects = sonarProjectRepository.
+                findSonarProjectsByProjectName(this.collector.getId(), this.instanceUrl, this.projectName);
+        return CollectionUtils.isNotEmpty(sonarProjects) ?
+                sonarProjects.stream().sorted(Comparator.comparing(SonarProject::getLastUpdated).reversed()).findFirst().get() : null;
+    }
+
+    private void updateCodeQualityData(SonarProject project) {
```

Comment:
@nameisaravind  can you move this code to DefaultClient, looks like it can be reused

Reply:
@nireeshT moved the method to util class, default client method is used in the method.